tests/unit/lms/views/api/canvas/sync_test.py

```diff
@@ -1,41 +1,250 @@
+    request_json,
 ):
     groupids = Sync(pyramid_request).sync()
 
     course_id = request_json["course"]["custom_canvas_course_id"]
     canvas_api_client.authenticated_users_sections.assert_called_once_with(course_id)
 
-    assert_sync_and_return(groupids, sections=sections.authenticated_user)
+    assert_sync_and_return_sections(groupids, sections=sections.authenticated_user)
+
+
+@pytest.mark.usefixtures("is_group_launch")
+def test_groups_sync(
+    pyramid_request,
+    assert_sync_and_return_groups,
+):
+
+    groups = [{"name": "group", "id": 1, "group_category_id": 2}]
+
+    with mock.patch.object(
+        Sync, "_get_canvas_groups", autospec=True
+    ) as _get_canvas_groups:
+
+        _get_canvas_groups.return_value = groups
+
+        groupids = Sync(pyramid_request).sync()
+
+    assert_sync_and_return_groups(groupids, groups=groups)
 
 
 @pytest.mark.usefixtures("user_is_instructor")
-def test_sync_when_the_user_is_an_instructor(
-    pyramid_request, canvas_api_client, sections, assert_sync_and_return, request_json
+def test_sections_sync_when_the_user_is_an_instructor(
+    pyramid_request,
+    canvas_api_client,
+    sections,
+    assert_sync_and_return_sections,
+    request_json,
 ):
     groupids = Sync(pyramid_request).sync()
 
     course_id = request_json["course"]["custom_canvas_course_id"]
     canvas_api_client.course_sections.assert_called_once_with(course_id)
 
-    assert_sync_and_return(groupids, sections=sections.course)
+    assert_sync_and_return_sections(groupids, sections=sections.course)
+
+
+@pytest.mark.usefixtures("user_is_learner", "is_group_launch")
+def test_get_canvas_groups_learner(pyramid_request):
+    # pylint: disable=protected-access
+    with mock.patch.object(
+        Sync, "_canvas_learner_group", autospec=True
+    ) as _canvas_learner_group:
+        groups = Sync(pyramid_request)._get_canvas_groups()
+
+        _canvas_learner_group.assert_called_once()
+        assert groups == [_canvas_learner_group.return_value]
+
+
+@pytest.mark.usefixtures("is_group_and_speedgrader", "user_is_instructor")
+def test_get_canvas_groups_speedgrader(pyramid_request):
+    # pylint: disable=protected-access
+    with mock.patch.object(
+        Sync, "_canvas_speedgrader_group", autospec=True
+    ) as _canvas_speedgrader_group:
+        groups = Sync(pyramid_request)._get_canvas_groups()
+
+        print(pyramid_request.json)
+
+        _canvas_speedgrader_group.assert_called_once()
+        assert groups == [_canvas_speedgrader_group.return_value]
+
+
+@pytest.mark.usefixtures("user_is_instructor")
+def test_get_canvas_groups_instructor(pyramid_request):
+    # pylint: disable=protected-access
+    with mock.patch.object(
+        Sync, "_canvas_course_groups", autospec=True
+    ) as _canvas_course_groups:
+        groups = Sync(pyramid_request)._get_canvas_groups()
+
+        _canvas_course_groups.assert_called_once()
+        assert groups == _canvas_course_groups.return_value
```

Comment:
The necessity of the `mock.patch.object()`'s might support the suggestion that these `_canvas_()` methods belong in the service